module/os/windows/spl/spl-kmem.c

```diff
@@ -0,0 +1,6628 @@
+	 * complete, purge the cache, and then destroy it.
+	 */
+	mutex_enter(&kmem_cache_lock);
+	list_remove(&kmem_caches, cp);
+	mutex_exit(&kmem_cache_lock);
+
+	if (kmem_taskq != NULL)
+		taskq_wait(kmem_taskq);
+
+	if (kmem_move_taskq != NULL && cp->cache_defrag != NULL)
+		taskq_wait(kmem_move_taskq);
+
+	kmem_cache_magazine_purge(cp);
+
+	mutex_enter(&cp->cache_lock);
+
+	if (cp->cache_buftotal != 0)
+		cmn_err(CE_WARN, "kmem_cache_destroy: '%s' (%p) not empty",
+		    cp->cache_name, (void *)cp);
+	if (cp->cache_defrag != NULL) {
+		avl_destroy(&cp->cache_defrag->kmd_moves_pending);
+		list_destroy(&cp->cache_defrag->kmd_deadlist);
+		kmem_cache_free(kmem_defrag_cache, cp->cache_defrag);
+		cp->cache_defrag = NULL;
+	}
+	/*
+	 * The cache is now dead.  There should be no further activity.  We
+	 * enforce this by setting land mines in the constructor, destructor,
+	 * reclaim, and move routines that induce a kernel text fault if
+	 * invoked.
+	 */
+#if defined(__aarch64__)
+	/* Setting landmines causes Break 0xC470: Ptrauth failure */
+	cp->cache_constructor = (int (*)(void *, void *, int))NULL;
+	cp->cache_destructor = (void (*)(void *, void *))NULL;
+	cp->cache_reclaim = (void (*)(void *))NULL;
+	cp->cache_move = (kmem_cbrc_t (*)(void *, void *, size_t, void *))NULL;
+#else
+	cp->cache_constructor = (int (*)(void *, void *, int))1;
+	cp->cache_destructor = (void (*)(void *, void *))2;
+	cp->cache_reclaim = (void (*)(void *))3;
+	cp->cache_move = (kmem_cbrc_t (*)(void *, void *, size_t, void *))4;
+#endif
+	mutex_exit(&cp->cache_lock);
+
+	kstat_delete(cp->cache_kstat);
+
+	if (cp->cache_hash_table != NULL)
+		vmem_free_impl(kmem_hash_arena, cp->cache_hash_table,
+		    (cp->cache_hash_mask + 1) * sizeof (void *));
+
+	for (cpu_seqid = 0; cpu_seqid < max_ncpus; cpu_seqid++)
+		mutex_destroy(&cp->cache_cpu[cpu_seqid].cc_lock);
+
+	mutex_destroy(&cp->cache_depot_lock);
+	mutex_destroy(&cp->cache_lock);
+
+	vmem_free_impl(kmem_cache_arena, cp, KMEM_CACHE_SIZE(max_ncpus));
+}
+
+static void
+kmem_alloc_caches_create(const int *array, size_t count,
+    kmem_cache_t **alloc_table, size_t maxbuf,
+    uint_t shift)
+{
+	char name[KMEM_CACHE_NAMELEN + 1];
+	size_t table_unit = (1 << shift); /* range of one alloc_table entry */
+	size_t size = table_unit;
+	int i;
+
+	for (i = 0; i < count; i++) {
+		size_t cache_size = array[i];
+		size_t align = KMEM_ALIGN;
+		kmem_cache_t *cp;
+
+		/* if the table has an entry for maxbuf, we're done */
+		if (size > maxbuf)
+			break;
+
+		/* cache size must be a multiple of the table unit */
+		ASSERT(P2PHASE(cache_size, table_unit) == 0);
+
+		/*
+		 * If they allocate a multiple of the coherency granularity,
+		 * they get a coherency-granularity-aligned address.
+		 */
+		if (IS_P2ALIGNED(cache_size, 64))
+			align = 64;
+		if (IS_P2ALIGNED(cache_size, PAGESIZE))
+			align = PAGESIZE;
+		(void) snprintf(name, sizeof (name),
+		    "kmem_alloc_%lu", cache_size);
```

Comment:
Provided that you take my previous suggestion to make PRI_SIZE, we can use that here.